service/src/main/java/datawave/microservice/query/mapreduce/MapReduceQueryJobRunner.java

```diff
@@ -0,0 +1,272 @@
+                    IllegalArgumentException e = new IllegalArgumentException(
+                                    "Invalid input format class specified. Must use one of " + mapReduceQueryProperties.getValidInputFormats());
+                    QueryException qe = new QueryException(DatawaveErrorCode.INVALID_FORMAT, e);
+                    log.error(qe.getMessage(), e);
+                    throw qe;
+                }
+            }
+            
+            try {
+                job.submit();
+            } catch (Exception e) {
+                QueryException qe = new QueryException(DatawaveErrorCode.MAPREDUCE_JOB_START_ERROR, e);
+                log.error(qe.getMessage(), qe);
+                throw qe;
+            }
+            
+            JobID mapReduceJobId = job.getJobID();
+            log.info("JOB ID: " + mapReduceJobId);
+            
+            // This is used to lookup the query id when the mapreduce callback comes in
+            // TODO: JWO: Should we use a CustomJobEndNotifier that passes the map reduce query id to the update state call instead?
+            mapReduceQueryCache.putQueryIdByJobIdLookup(mapReduceJobId.toString(), mapReduceQueryStatus.getId());
+            
+            mapReduceQueryStatus.setJobId(mapReduceJobId.toString());
+            mapReduceQueryStatus.setJobTracker(mapReduceJob.getMapReduceJobProperties().getJobTracker());
+            mapReduceQueryStatus.setHdfsUri(mapReduceJob.getMapReduceJobProperties().getHdfsUri());
+            mapReduceQueryStatus.setState(MapReduceQueryStatus.MapReduceQueryState.SUBMITTED);
+        } catch (QueryException e) {
+            mapReduceQueryStatus.setErrorCode(DatawaveErrorCode.findCode(e.getErrorCode()));
+            mapReduceQueryStatus.setFailureMessage(e.getMessage());
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            PrintWriter writer = new PrintWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
+            e.printStackTrace(writer);
+            writer.close();
+            mapReduceQueryStatus.setState(MapReduceQueryStatus.MapReduceQueryState.FAILED);
+            mapReduceQueryStatus.setStackTrace(new String(outputStream.toByteArray(), StandardCharsets.UTF_8));
+        } catch (Exception e) {
+            mapReduceQueryStatus.setErrorCode(DatawaveErrorCode.UNKNOWN_SERVER_ERROR);
+            mapReduceQueryStatus.setFailureMessage(e.getMessage());
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+            PrintWriter writer = new PrintWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
+            e.printStackTrace(writer);
+            writer.close();
+            mapReduceQueryStatus.setState(MapReduceQueryStatus.MapReduceQueryState.FAILED);
+            mapReduceQueryStatus.setStackTrace(new String(outputStream.toByteArray(), StandardCharsets.UTF_8));
```

Comment:
Other than the setErrorCode, this appears to be copied code, also used below with the oozie code path  Maybe it's worth adding a function taking an error code

Reply:
i condensed the two exception blocks.